GitHub Actions Test Infrastructure

aditigopalan opened the pull request:
Fixes #13, Fixes #8 and Fixes #34

# Add nf-test Infrastructure

## Changes Made

1. **Test Framework Setup**
   - Added nf-test framework for automated testing
   - Created `nf-test.config` with basic configuration
   - Set up test fixtures in `tests/fixtures/` directory

2. **Initial Test Cases**
   - Added test for successful repository processing
   - Added test for invalid repository URL handling
 
3. **Parameter Validation Fixes**
   - Fixed validation order for Synapse parameters
   - Added validation for mutually exclusive parameters (sample_sheet and repo_url)
   - Improved error messages for parameter validation
   
## Testing
- Tests verify both success and error cases
- Test infrastructure is ready for adding more test cases

aditigopalan committed:
Adding test infra

aditigopalan committed:
Adding examples

aditigopalan committed:
Modify test.yml for param validation

aditigopalan committed:
Fix param validation in main.nf

aditigopalan committed:
Updating test location

aditigopalan committed:
Update test.yml

aditigopalan committed:
Update test.yml

aditigopalan committed:
Update test.yml

aditigopalan committed:
URL fix

aditigopalan committed:
Sample sheet fix

aditigopalan committed:
URL fix

aditigopalan committed:
Adding nf test

aditigopalan committed:
Update main.nf

aditigopalan committed:
Creating invalid-input.csv

For nf-test

aditigopalan committed:
Update test.yml

aditigopalan committed:
Fixing path for nf-test

aditigopalan committed:
nf-test fix

aditigopalan committed:
test-nf asserts

aditigopalan committed:
Update test.yml

aditigopalan committed:
nf-test config

aditigopalan committed:
Update test.yml

aditigopalan committed:
workflow > pipeline fix

aditigopalan committed:
Fix

aditigopalan committed:
nf-test fix

aditigopalan committed:
Create nextflow.config

aditigopalan committed:
Removing trace

aditigopalan committed:
Fix generate report

aditigopalan committed:
Delete nextflow.config (redundant)

aditigopalan committed:
Update nf-test.config

aditigopalan committed:
Updating configFile to null

BWMac commented:
@aditigopalan Has this workflow been test on Seqera Platform? If so, can you provide links to the run(s)?

Also, you should delete those `.DS_Store` files and add them to your `gitignore` file.

BWMac commented on main.nf.test:
```diff
@@ -0,0 +1,36 @@
+    name "CCKP Toolkit"
+    script "main.nf"
+
+    test("Should process a single repository URL") {
+        when {
+            params {
+                repo_url = "https://github.com/PythonOT/POT.git"
+                output_dir = "test_results"
+            }
+        }
+
+        then {
+            assert workflow.success
+            assert workflow.trace.tasks().size() > 0
+            assert workflow.trace.succeeded().size() > 0
+            assert workflow.trace.failed().size() == 0
+        }
+    }
```
Should there also be a test for the sample sheet passing scenario?

aditigopalan replied:
Yes! It seemed tricky to implement with nf-test, but I was looking at [nf-schema ](https://nextflow-io.github.io/nf-schema/latest/nextflow_schema/sample_sheet_schema_specification/) for sample sheet validation and will probably add that functionality once this is merged

BWMac commented on modules/GenerateReport.nf:
```diff
@@ -1,3 +1,6 @@
+#!/usr/bin/env nextflow
+nextflow.enable.dsl=2
```
I think you only need this line on `main.nf`, or was this causing issues with the test runs?

aditigopalan replied:
I did have problems while testing! Which is why I added it everywhere -- is this a common issue?

BWMac replied:
Interesting. I have never encountered that issue, but if I guess it isn't a big deal to include these. Out of curiosity, can you share the error you get with the test framework when you don't set `nextflow.enable.dsl=2` everywhere?

aditigopalan replied:
Weird, this time I tried removing it from the module files and it seems to work! Thanks for the catch

BWMac commented on modules/RunAlmanack.nf:
```diff
@@ -24,7 +27,7 @@
 
 process RunAlmanack {
-    container = 'python:3.11-slim'
+    container = 'python:3.11'
```
```suggestion
    container 'python:3.11'
```

BWMac commented on modules/RunAlmanack.nf:
```diff
@@ -46,7 +49,7 @@ process RunAlmanack {
 
     # Install git and pip
-    apt-get update && apt-get install -y git
+    apt-get update && apt-get install -y --no-install-recommends git && rm -rf /var/lib/apt/lists/*
 
```
Is there a hosted container image that has the python version you need and `git`? I think you typically want to avoid installing things in a Nextflow process.

BWMac commented on modules/UploadToSynapse.nf:
```diff
@@ -1,35 +1,46 @@
  */
 
 process UploadToSynapse {
+    container = 'python:3.11-slim'
```
```suggestion
    container 'python:3.11-slim'
```

BWMac commented on modules/UploadToSynapse.nf:
```diff
@@ -1,35 +1,46 @@
+ * 3. Uploads all output files
  */
 
 process UploadToSynapse {
```
This process should be using the `synapseclient` docker container
https://github.com/Sage-Bionetworks/synapsePythonClient/pkgs/container/synapsepythonclient

BWMac commented on modules/UploadToSynapse.nf:
```diff
@@ -1,35 +1,46 @@
+    # Authenticate with Synapse
+    synapse login -u ${params.synapse_username} -p ${params.synapse_password}
+
+    # Create a new folder for the results
+    synapse create -parentid ${params.synapse_folder_id} -name "${repo_name}_results"
+
+    # Upload all output files
+    for f in ${out_dir}/*; do
+        synapse store "\$f" --parentid ${params.synapse_folder_id}
+    done
+
```
Because this usecase involves potentially uploading many files at once, you might consider using an executable Python script here.

aditigopalan replied:
Logged this as an issue!

BWMac commented on modules/GenerateReport.nf:
```diff
@@ -15,19 +18,26 @@ process GenerateReport {
     
     script:
     """
+    #!/bin/bash
+    set -euo pipefail
```
Can you explain why this command is needed?

aditigopalan replied:
`set -euo pipefail`
This is a set of shell options that make the script more robust and safer:
`-e`: Exit immediately if any command returns a non-zero status (error)
`-u`: Treat unset variables as errors and exit immediately
`-o pipefail`: Return the exit status of the last command in a pipeline that failed, rather than the last command overall

Added these in the workflow because:
1. We want to fail fast if any command fails (like if Almanack analysis fails)
2. We want to catch any undefined variables that might cause issues
3. We want to ensure pipeline failures (like | commands) are properly detected

Trying to avoid:
1. Continue running even after critical errors
2. Silently use empty values for undefined variables
3. Miss failures in pipeline commands

BWMac replied:
Awesome! Thanks

BWMac reviewed (changes requested):
Awesome work! I left some comments/questions to think about. 

Also, when I was looking at your `nextflow.config` file I noticed a few things:
- Your process containers haven't been pinned to specific versions
- You have a `platform` parameter set in the `docker` section. AFAIK, that isn't a supported config option.

aditigopalan committed:
Update modules/RunAlmanack.nf

Co-authored-by: Brad Macdonald <52762200+BWMac@users.noreply.github.com>

aditigopalan committed:
Update modules/UploadToSynapse.nf

Co-authored-by: Brad Macdonald <52762200+BWMac@users.noreply.github.com>

aditigopalan committed:
Add warn for invalid urls on sample sheet

aditigopalan committed:
refacrefactor: improve error handling and parameter validation

aditigopalan committed:
Adding IllegalArgumentException

aditigopalan committed:
Merge branch 'dev' of https://github.com/mc2-center/cckp-toolkit-workflow into dev

aditigopalan committed:
Fix

aditigopalan committed:
test: add validation tests for sample sheet and missing parameters

- Add test for missing repo_url column in sample sheet
- Add test for missing both sample_sheet and repo_url parameters
- Fix error message format to match Nextflow's output
- Improve sample sheet validation to prevent null pointer exceptions

aditigopalan committed:
test: add test for sample sheet processing

- Add test for successful processing of multiple repositories from sample sheet
- Fix channel handling in sample sheet processing
- Remove redundant channel assignment

aditigopalan committed:
Remove git install command

Python:3.11 has git installed already

aditigopalan committed:
Update UploadtoSynapse container

aditigopalan committed:
Adding invalid-sample-sheet.csv

aditigopalan committed:
Remove enable dsl 2 from module files

aditigopalan committed:
Config fixes

aditigopalan commented:
> Awesome work! I left some comments/questions to think about.
> 
> Also, when I was looking at your `nextflow.config` file I noticed a few things:
> 
> * Your process containers haven't been pinned to specific versions
> * You have a `platform` parameter set in the `docker` section. AFAIK, that isn't a supported config option.

@BWMac thank you so much for catching these! I fixed your suggested changes, let me know what you think  🚀

BWMac commented on .nf-test/tests/a5db341bb7d3e73dbb7674a87b560ca/meta/nextflow.log:
```diff
@@ -0,0 +1,74 @@
+Mar-14 11:18:08.369 [main] DEBUG nextflow.cli.Launcher - $> nextflow -quiet -log /Users/agopalan/cckp-toolkit-workflow/.nf-test/tests/a5db341bb7d3e73dbb7674a87b560ca/meta/nextflow.log run /Users/agopalan/cckp-toolkit-workflow/main.nf -c /Users/agopalan/cckp-toolkit-workflow/nextflow.config -params-file /Users/agopalan/cckp-toolkit-workflow/.nf-test/tests/a5db341bb7d3e73dbb7674a87b560ca/meta/param ...
```
make sure you delete this and add the folders/files that nextflow/nf-test generates to your gitignore

BWMac commented on modules/UploadToSynapse.nf:
```diff
@@ -1,35 +1,42 @@
+    set -euxo pipefail
+
+    # Authenticate with Synapse
+    synapse login -u ${params.synapse_username} -p ${params.synapse_password}
```
I thought I commented on this before, but apparently I didn't (my bad). This line should be replaced with a Nextflow [secret](https://www.nextflow.io/docs/latest/secrets.html). You can configure secrets both with your local installation of Nextflow and in your Seqera Platform workspace. When running the workflow in SP, you will need to provide the workflow secret when configuring the run. If you create a secret `SYNAPSE_AUTH_TOKEN` your auth token becomes an env variable within the process machine, and the Python client will pick up on it automatically so there will be no need for a `synapse login` line.

Example usage in `nf-synapse` https://github.com/Sage-Bionetworks-Workflows/nf-synapse/blob/92ce850a396b0ccd1045eb19f6031372d440ae0d/modules/synapse_get.nf#L8

aditigopalan replied:
Thank you, I have a secret on the tower but haven't configured it here yet. I'll add this change in another branch along with some additional modifications (I might remove the upload to synapse function for external users)

BWMac replied:
I recommend doing this now rather than later. Your auth token will be on display in the SP UI as this is currently written because you have to pass it as a parameter.

aditigopalan replied:
That makes sense, thanks!

BWMac reviewed (commented):
Great changes! Just a couple more comments.

aditigopalan committed:
Update .gitignore

aditigopalan committed:
Delete .nf-test/tests directory

aditigopalan committed:
add synapse auth token as secret

BWMac reviewed (approved):
🔥 🔥 🔥 🔥 🔥 🔥 🔥
